### PR TITLE
Generate supplicant ready event

### PR DESCRIPTION
**Pull request by rado17**
Generate supplicant ready event once the control connection between cli and supplicant is setup.

### CODE REVIEW

File: zephyr/src/supp_events.h [hunk 1]
```diff
@@ -3,6 +3,35 @@
 
 #include <zephyr/net/wifi_mgmt.h>
 
+/* Connectivity Events */
+#define _NET_MGMT_WPA_SUPP_LAYER			NET_MGMT_LAYER_L2
+#define _NET_MGMT_WPA_SUPP_CODE	            0x157
+#define _NET_MGMT_WPA_SUPP_BASE			(NET_MGMT_LAYER(_NET_MGMT_WPA_SUPP_LAYER) \
+										| NET_MGMT_LAYER_CODE(_NET_MGMT_WPA_SUPP_CODE)  | \
+											NET_MGMT_IFACE_BIT)
+#define _NET_MGMT_WPA_SUPP_EVENT			(NET_MGMT_EVENT_BIT | _NET_MGMT_WPA_SUPP_BASE)
```
Comment (krish2718) [inline zephyr/src/supp_events.h]: alignment looks off.

File: zephyr/src/supp_main.c [hunk 2]
```diff
@@ -162,7 +165,12 @@ static int z_wpas_add_interface(const char* ifname)
 		k_mutex_unlock(&iface_up_mutex);
 	}
 
-	z_wpa_ctrl_init(wpa_s);
+	ret = z_wpa_ctrl_init(wpa_s);
+	if (ret) {
+		wpa_printf(MSG_ERROR, "Failed to initialize control interface");
+		return ret;
+	}
+	send_wifi_state_event(ifname, NET_EVENT_WPA_SUPP_CMD_READY);
```
Comment (krish2718) [inline zephyr/src/supp_main.c]: This is sent for each interface addition, it should only be sent for 1st interface add.

File: zephyr/src/supp_main.c [hunk 3]
```diff
@@ -214,6 +222,7 @@ static int z_wpas_remove_interface(const char* ifname)
 		return -1;
 	}
 
+	send_wifi_state_event(ifname, NET_EVENT_WPA_SUPP_CMD_IFACE_REMOVED);
```
Comment (krish2718) [inline zephyr/src/supp_main.c]: If all interfaces are removed we should also send `NET_EVENT_WPA_SUPP_CMD_NOT_READY`

File: zephyr/src/supp_events.c [hunk 4]
```diff
@@ -27,3 +27,31 @@ int send_wifi_mgmt_event(const char *ifname, enum net_event_wifi_cmd event, int
...
+	const struct device *dev = device_get_binding(ifname);
+	struct net_if *iface = net_if_lookup_by_dev(dev);
+
+	if (!iface) {
+		wpa_printf(MSG_ERROR, "Could not find iface for %s", ifname);
+		return -ENODEV;
+	}
+
+	switch (event) {
+	case NET_EVENT_WPA_SUPP_CMD_READY:
```
Comment (krish2718) [inline zephyr/src/supp_events.c:48]: NET_EVENT_WPA_SUPP_CMD_NOT_READY is not implemented

File: zephyr/src/supp_events.c [hunk 5]
```diff
@@ -27,3 +27,31 @@ int send_wifi_mgmt_event(const char *ifname, enum net_event_wifi_cmd event, int
...
+		net_mgmt_event_notify(NET_EVENT_WPA_SUPP_READY, iface);
+		break;
+	case NET_EVENT_WPA_SUPP_CMD_IFACE_ADDED:
+		net_mgmt_event_notify(NET_EVENT_WPA_SUPP_IFACE_ADDED, iface);
+		break;
+	case NET_EVENT_WPA_SUPP_CMD_IFACE_REMOVED:
+		net_mgmt_event_notify(NET_EVENT_WPA_SUPP_IFACE_REMOVED, iface);
```
Comment (krish2718) [inline zephyr/src/supp_events.c]: can we assume `iface` will be valid here?

File: zephyr/src/supp_events.c [hunk 6]
```diff
@@ -27,3 +27,31 @@ int send_wifi_mgmt_event(const char *ifname, enum net_event_wifi_cmd event, int
 
 	return 0;
 }
+
+int send_wifi_state_event(const char *ifname, enum net_event_wpa_supp_cmd event)
+{
+	const struct device *dev = device_get_binding(ifname);
```
Comment (krish2718) [inline zephyr/src/supp_events.c:34]: Add a TODO here, this is deprecated and needs to be removed soon.

File: zephyr/src/supp_main.c [hunk 7]
```diff
@@ -447,6 +456,8 @@ static void z_wpas_start(void)
 	close(z_wpas_event_sockpair[0]);
 	close(z_wpas_event_sockpair[1]);
 
+	send_wifi_state_event(DEFAULT_IFACE_NAME, NET_EVENT_WPA_SUPP_CMD_NOT_READY);
```
Comment (krish2718) [inline zephyr/src/supp_main.c]: Shouldn't we send this before we start de-initializing supplicant.

File: zephyr/src/supp_main.c [hunk 8]
```diff
@@ -154,6 +155,8 @@ static int z_wpas_add_interface(const char* ifname)
 		wpa_printf(MSG_ERROR, "Failed to add iface: %s", ifname);
 		return -1;
 	}
+
+	send_wifi_state_event(ifname, NET_EVENT_WPA_SUPP_CMD_IFACE_ADDED);
```
Comment (krish2718) [inline zephyr/src/supp_main.c]: This should be after control interface init to avoid command failures.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: zephyr/src/supp_events.c (hunk 5 above)
Comment (rado17) [reply zephyr/src/supp_events.c]: We are checking for iface pointer validity before the switch case.

Comment (krish2718) [reply zephyr/src/supp_events.c]: Sure, but its more than checking the validity, if its always NULL then this event is useless.

File: zephyr/src/supp_events.h (hunk 1 above)
Comment (rado17) [reply zephyr/src/supp_events.h]: Corrected.

File: zephyr/src/supp_main.c (hunk 2 above)
Comment (rado17) [reply zephyr/src/supp_main.c]: Done

File: zephyr/src/supp_main.c (hunk 3 above)
Comment (rado17) [reply zephyr/src/supp_main.c]: Done

File: zephyr/src/supp_events.c (hunk 4 above)
Comment (rado17) [reply zephyr/src/supp_events.c:48]: Done

File: zephyr/src/supp_events.c (hunk 5 above)
Comment (rado17) [reply zephyr/src/supp_events.c]: Added dev validity check before fetching iface.

File: zephyr/src/supp_main.c (hunk 7 above)
Comment (rado17) [reply zephyr/src/supp_main.c]: Done.

File: zephyr/src/supp_main.c (hunk 8 above)
Comment (rado17) [reply zephyr/src/supp_main.c]: Done

File: zephyr/src/supp_events.c (hunk 6 above)
Comment (rado17) [reply zephyr/src/supp_events.c:34]: Done

File: zephyr/src/supp_main.c [hunk 9]
```diff
@@ -214,6 +226,12 @@ static int z_wpas_remove_interface(const char* ifname)
 		return -1;
 	}
 
+	if (z_wpas_get_iface_count() == 0) {
+		send_wifi_state_event(DEFAULT_IFACE_NAME, NET_EVENT_WPA_SUPP_CMD_NOT_READY);
+	}
+
+	send_wifi_state_event(ifname, NET_EVENT_WPA_SUPP_CMD_IFACE_REMOVED);
```
Comment (krish2718) [inline zephyr/src/supp_main.c]: This should be before sending not_ready, and I am also thinking if this should be send before we actually remove the interface? E.g., 

1. Removed interface
2. Application still send request on that iface
3. IFACE_REMOVED is sent

So, what will happen to #2, do we gracefully handle it? I guess even if we move #3 before #1 there could still be a race, but it narrows it.

Comment (rado17) [reply zephyr/src/supp_main.c]: True. But I placed this at the end considering the case where we send the event but interface removal fails.

Comment (krish2718) [reply zephyr/src/supp_main.c]: Well, need to handle both, so, may be, use two events, `REMOVING` and then `REMOVED`?

Comment (rado17) [reply zephyr/src/supp_main.c]: Done

**Comment by krish2718** [comment]
Please close this as the main one is merged.

**Pull request closed without merging**